fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/AttachmentContainerImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.attachment;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.nbt.NbtElement;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.attachment.v1.AttachmentContainer;
+import net.fabricmc.fabric.api.attachment.v1.AttachmentType;
+
+@ApiStatus.Internal
+public class AttachmentContainerImpl implements AttachmentContainer {
+	public static final String CONTAINER_KEY = "fabric:attachments";
+
+	private final Map<AttachmentType<?>, Object> attachments = new IdentityHashMap<>();
+
+	@Override
+	@Nullable
+	public <T> T get(AttachmentType<T> attachment) {
+		Objects.requireNonNull(attachment, "Attachment type may not be null.");
+
+		return (T) attachments.get(attachment);
+	}
+
+	@Override
+	@Nullable
+	public <T> T set(AttachmentType<T> attachment, T value) {
+		Objects.requireNonNull(attachment, "Attachment type may not be null.");
+		Objects.requireNonNull(value, "Value may not be null.");
+
+		return (T) attachments.put(attachment, value);
+	}
+
+	@Override
+	@Nullable
+	public <T> T remove(AttachmentType<T> attachment) {
+		Objects.requireNonNull(attachment, "Attachment type may not be null.");
+
+		return (T) attachments.remove(attachment);
+	}
+
+	@SuppressWarnings({"rawtypes", "unchecked"})
+	@Override
+	public void writeNbt(NbtCompound nbt, String subtag) {
+		NbtCompound attachmentsTag = new NbtCompound();
+
+		for (Map.Entry<AttachmentType<?>, Object> entry : attachments.entrySet()) {
```

Comment:
use forEach lambda; identity hash map does not store by entry, so iterating entry set allocates extra objects

Reply:
In the current version of the PR I don't think I can use forEach without introducing a capturing lambda.